### PR TITLE
Intent detection/slot tagging/domain detection

### PR DESCRIPTION
**Pull request by VahidooX**
+ Added slot masking (specifies which outputs are considered in the slot tagging loss) and real token masking (specified which sub-tokens represent the predictions) to the outputs of data layers.
+ Passing slot masking instead of input masking to the loss function for slot tagging.
+ Fixed a bug in the inference for batch size=1 for cases where a word is tokenized into multiple sub-tokens (New masking outputs seem necessary to fix this bug).
+ Printing detail of slot tagging performance during training in the output
+ Made the evaluation processes during training and evaluation consistent
+ Added importing functions to process and import triple datasets
+ Some updates to the list of the sub-tokens considered in the slot tagging loss

### CODE REVIEW

File: collections/nemo_nlp/nemo_nlp/data/datasets/joint_intent_slot.py [hunk 1]
```diff
@@ -39,37 +39,47 @@ def get_features(queries,
...
             slots = [pad_label]
 
         for j, word in enumerate(words):
             word_tokens = tokenizer.tokenize(word)
             subtokens.extend(word_tokens)
-            slot_mask.append(True)
-            slot_mask.extend([False] * (len(word_tokens) - 1))
+
+            loss_mask.append(True)
+            loss_mask.extend([True] * (len(word_tokens) - 1))
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/joint_intent_slot.py]: should we add an option for users to choose whether to take into the loss of non-beginning tokens? If that option exists, then we should process loss_mask accordingly

File: collections/nemo_nlp/nemo_nlp/data/data_layers.py [hunk 2]
```diff
@@ -131,7 +131,11 @@ def create_ports():
                 0: AxisType(BatchTag),
                 1: AxisType(TimeTag)
             }),
-            "token_mask": NeuralType({
+            "loss_mask": NeuralType({
+                0: AxisType(BatchTag),
+                1: AxisType(TimeTag)
+            }),
+            "subtokens_mask": NeuralType({
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/data_layers.py]: please modify the docstring to explain the different between loss_mask and subtokens_mask

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 3]
```diff
@@ -312,6 +317,106 @@ def process_atis(infold, uncased, modes=['train', 'test'], dev_split=0):
...
+                                           int(end_i),
+                                           slot_name])
+                    if slot_name not in slots_list:
+                        slots_list[slot_name] = len(slots_list)
+                        slots_list_all[f'B-{slot_name}'] = len(slots_list_all)
+                        slots_list_all[f'I-{slot_name}'] = len(slots_list_all)
+                        outfiles['dict_slots'].write(f'B-{slot_name}\n')
+                        outfiles['dict_slots'].write(f'I-{slot_name}\n')
+            else:
+                slot_tags_list = []
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: put slot_tags_list = [] outside the if/else

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 4]
```diff
@@ -312,6 +317,106 @@ def process_atis(infold, uncased, modes=['train', 'test'], dev_split=0):
...
+                outfiles['dict_intents'].write(f'{intent_str}\n')
+
+            outfiles[mode].write(f'{sentence[4:-4]}\t'
+                                 f'{str(intents_list[intent_str])}\n')
+
+            if str.count(slot_tags_str, ":") > 0:
+                slot_tags = slot_tags_str.strip().split(",")
+                slot_tags_list = []
+                for st in slot_tags:
+                    if len(st.strip()) == 0:
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: if not str.strip():

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 5]
```diff
@@ -312,6 +317,106 @@ def process_atis(infold, uncased, modes=['train', 'test'], dev_split=0):
...
+            [intent_str, slot_tags_str, sentence] = query.strip().split("\t")
+
+            if intent_str not in intents_list:
+                intents_list[intent_str] = len(intents_list)
+                outfiles['dict_intents'].write(f'{intent_str}\n')
+
+            outfiles[mode].write(f'{sentence[4:-4]}\t'
+                                 f'{str(intents_list[intent_str])}\n')
+
+            if str.count(slot_tags_str, ":") > 0:
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: do we need to count ":"? it looks like if a sentence has no slot tag, it's just empty, so maybe check if slot_tags_str.strip() isn't empty is enough?

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 6]
```diff
@@ -312,6 +317,106 @@ def process_atis(infold, uncased, modes=['train', 'test'], dev_split=0):
     return outfold
 
 
+def generate_BIO(slot_tags_list, token_str, start_i, end_i):
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: I don't think we need this function, but if we decide to keep this, there are a couple of things:

1. please add docstring to explain why we need this.
2. can end_i ever be greater than tag_end?

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 7]
```diff
@@ -312,6 +317,106 @@ def process_atis(infold, uncased, modes=['train', 'test'], dev_split=0):
...
+                    if slot_name not in slots_list:
+                        slots_list[slot_name] = len(slots_list)
+                        slots_list_all[f'B-{slot_name}'] = len(slots_list_all)
+                        slots_list_all[f'I-{slot_name}'] = len(slots_list_all)
+                        outfiles['dict_slots'].write(f'B-{slot_name}\n')
+                        outfiles['dict_slots'].write(f'I-{slot_name}\n')
+            else:
+                slot_tags_list = []
+
+            span_generator = WhitespaceTokenizer().span_tokenize(sentence)
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: I think this entire function can be rewritten to be a lot more efficient and easier to understand. So the slots are indexed like this: 
start0:end0:label0, start1:end1:label1, ...

So for each sentence, we can do this:
- start at index 0. all words from index 0 to index start0 is tagged with 'O'
- for each chunk starti:endi: the first word is add B-tag, the rest is added I-tag
- for each chunk between endi:starti+1, each word is labeled with 'O'

Comment (chiphuyen) [reply collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: I can also merge your PR as is and rewrite the function

File: collections/nemo_nlp/nemo_nlp/data/datasets/joint_intent_slot.py (hunk 1 above)
Comment (VahidooX) [reply collections/nemo_nlp/nemo_nlp/data/datasets/joint_intent_slot.py]: Added two parameters: 1-mask the extra subtokens, and 2-[CLS] and [SEP] tokens.

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py (hunk 6 above)
Comment (VahidooX) [reply collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: Merged it into the caller function. end_i can be larger in cases that start_i is also larger than the tag_start.

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py (hunk 7 above)
Comment (VahidooX) [reply collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: Seems a good idea. I can rewrite it later more efficiently without WhitespaceTokenizer.
